### PR TITLE
The initial stub at adding support for cAdvisor

### PR DESCRIPTION
**Pull request by ichekrygin**
This PR adds the initial support for metrics collected from the Kubelet's advisor sub-service, with the initial support for Network metrics only.

### CODE REVIEW

File: pkg/sampler/cache.go [hunk 1]
```diff
@@ -162,6 +171,54 @@ func (s *sampleCache) getAllSamples() (allSampleInstants, int) {
...
+		PodUID:        podUID,
+		ContainerName: containerName,
+	}
+}
+
+func (s *sampleCache) fetchCAdvisorSample(samples *sampleInstants) error {
+	if samples == nil {
+		return nil
+	}
+	containers, err := s.CadvisorClient.SubcontainersInfo("kubelet", &v1.ContainerInfoRequest{})
```
Comment (campoy) [inline pkg/sampler/cache.go:193]: Is the kubelet the highest level in the container hierarchy? Does this include system and kubepod workloads?

File: pkg/sampler/cache.go [hunk 2]
```diff
@@ -162,6 +171,54 @@ func (s *sampleCache) getAllSamples() (allSampleInstants, int) {
...
+		// Assert pod containers.
+		key := ContainerKeyFromContainerInfo(container)
+		if key.PodUID == "" {
+			continue // Skip all non-pod containers.
+		}
+
+		// Assert containers for current samples.
+		sample, found := samples.containers[key]
+		if !found {
+			// Skip containers that are not found in the provided samples.
```
Comment (campoy) [inline pkg/sampler/cache.go:208]: Might be worth logging these if it's something we don't expect to see

File: pkg/sampler/types.go [hunk 3]
```diff
@@ -69,6 +71,12 @@ type sampleInstant struct {
 
 	// MemoryUsageLifetimeMaxBytes uint64
 	// MemoryLimitBytes            uint64
+
+	// CAdvisor stats.
+	// Although, currently, we only collect and report network metrics,
+	// we [re]use ContainerStats as sample value holder since in the future
+	// we may chose to collect addition values (other than Network)
```
Comment (campoy) [inline pkg/sampler/types.go:78]:
```suggestion
	// we may chose to collect additional values (other than Network)
```

File: pkg/sampler/cache.go [hunk 4]
```diff
@@ -323,6 +387,9 @@ func (s *sampleCache) metricToSample(
 	return sample
 }
 
+// cgroups resource file was not updated between two consecutive fetches.
```
Comment (campoy) [inline pkg/sampler/cache.go:390]: I'm not sure I understand this comment.

File: proto/pkg/cadvisor/api/cadvisor.proto [hunk 5]
```diff
@@ -0,0 +1,171 @@
+syntax = "proto3";
+
+package cadvisor.api;
+option go_package = "./api";
+import "google/protobuf/timestamp.proto";
+import "google/api/annotations.proto";
+
+service ContainerInfoService {
```
Comment (campoy) [inline proto/pkg/cadvisor/api/cadvisor.proto:8]: Do we need this service? Also, why is the rpc named `NodeCapacityStatus`?

File: pkg/sampler/cache.go (hunk 1 above)
Comment (ichekrygin) [reply pkg/sampler/cache.go:193]: That's a good question. I am not sure of the top.

File: pkg/sampler/cache.go (hunk 4 above)
Comment (ichekrygin) [reply pkg/sampler/cache.go:390]: I don't recognize this comment. It shouldn't be part of this PR, I will double-check.

Comment (ichekrygin) [reply pkg/sampler/cache.go]: Ah, yes. Removed.

File: proto/pkg/cadvisor/api/cadvisor.proto (hunk 5 above)
Comment (ichekrygin) [reply proto/pkg/cadvisor/api/cadvisor.proto]: Unintended commit. Removed.

File: pkg/sampler/cache.go (hunk 2 above)
Comment (ichekrygin) [reply pkg/sampler/cache.go:211]: Done.

**Review by campoy — APPROVED** [review]

File: pkg/sampler/cache.go (hunk 1 above)
Comment (pwittrock) [reply pkg/sampler/cache.go]: I think we just empty string for the cgroup for the top level.

**Pull request merged**